examples/permissioned-token/assets/permissions.py

```diff
@@ -0,0 +1,156 @@
+from pyteal import *
+
+def approval_program():
+    """
+    Th Permissions smart contract defines transfer rules.
+    Here, we implement two rules:
+        a) receiver asset balance should be <=100 after transfer
+        b) both [from, to] accounts must be whitelisted
+    The Permissions smart contract must be called with the Controller to validate
+    transfers. The Controller smart contract defines a manager account, which
+    is used here to update user attributes and smart contract state.
+    """
+
+    true = Int(1)
+    max_tokens = Bytes("max_tokens") # maximum token transfer rule
+    whitelist_count = Bytes("whitelist_count") # whitelisted accounts counter
+    controller_app_id = Bytes("controller_app_id") # token's controller application index
+
+    # Always verify that the RekeyTo property of any transaction is set to the ZeroAddress
+    # unless the contract is specifically involved ina rekeying operation.
+    rekey_check = Txn.rekey_to() == Global.zero_address()
+
+    """
+    Handles Permissions SSC deployment. Expectes 1 argument
+    * controller_app_id: controller application index
+    During deployment
+    * max_tokens is set to 100
+    * whitelist_counter is intialized to 0
+    * controller application index is saved in global state
+    """
+    on_deployment = Seq([
+    	Assert(And(
+    		Txn.application_args.length() == Int(1),
+            rekey_check
+    	)),
+
+        # Save max_tokens count in global state (default = 100, during ssc deploy)
+        App.globalPut(max_tokens, Int(100)),
+
+        # Initialize whitelisted accounts counter to 0
+        App.globalPut(whitelist_count, Int(0)),
+
+        # Save controller application index in global state
+        App.globalPut(controller_app_id, Btoi(Txn.application_args[0])),
+        Return(Int(1))
+    ])
+
+    # fetch permissions manager, token_id from controller's global state (using foreignApps)
+    permissions_manager = App.globalGetEx(Int(1), Bytes("manager"))
+    token_id = App.globalGetEx(Int(1), Bytes("token_id"))
+
+    """
+    Whitelist an account. If a non-reserve account is whitelisted then it can receive tokens.
+    Only accepted if txn sender is the permissions manager.
+    1 expected argument:
+    * controller_app_id: controller application index
+    NOTE: controller_app_id is also passed in txn.ForeignApps (to load perm_id, manager from controller's global state)
+    """
+    add_whitelist = Seq([
+        permissions_manager,
+        token_id,
+        Assert(And(
+    		Txn.application_args.length() == Int(2),
+    		Txn.accounts.length() == Int(1),
+    		rekey_check,
+    		# Txn.applications.length() ==  Int(1), [TEALv3]
+
+    		# verify from global state whether controller application passed is the valid one
+    		# note: with tealv3 we can just use txn.applications[0] instead of passing an an appArg
+    		Btoi(Txn.application_args[1]) == App.globalGet(controller_app_id),
```

Comment:
why do we need to verify a controller_app_id here? I think this "branch" can be called without calling controller.

Reply:
We need this check because we fetch *permissions_manager* from controller's global state. And `Txn.sender()` should be permissions_manager

Reply:
ah, correct. Please add this explanation to the comment in the code.